sw/device/tests/sim_dv/spi_tpm_smoketest.c

```diff
@@ -0,0 +1,217 @@
+                               &spi_device_irq);
+
+  switch (spi_device_irq) {
+    case kDifSpiDeviceIrqTpmHeaderNotEmpty:
+      header_interrupt_received = true;
+      // Disable interrupt until work is handled.
+      CHECK_DIF_OK(dif_spi_device_irq_set_enabled(
+          &spi_device.dev, kDifSpiDeviceIrqTpmHeaderNotEmpty,
+          kDifToggleDisabled));
+      break;
+    default:
+      LOG_ERROR("Unexpected interrupt: %d", spi_device_irq);
+      break;
+  }
+}
+
+static void ack_spi_tpm_header_irq(dif_spi_device_handle_t *spi_device) {
+  // Clear interrupt state and re-enable interrupt.
+  CHECK_DIF_OK(dif_spi_device_irq_acknowledge(
+      &spi_device->dev, kDifSpiDeviceIrqTpmHeaderNotEmpty));
+  CHECK_DIF_OK(dif_spi_device_irq_set_enabled(
+      &spi_device->dev, kDifSpiDeviceIrqTpmHeaderNotEmpty, kDifToggleEnabled));
+}
+
+bool test_main(void) {
+  CHECK_DIF_OK(dif_pinmux_init(
+      mmio_region_from_addr(TOP_EARLGREY_PINMUX_AON_BASE_ADDR), &pinmux));
+
+  CHECK_DIF_OK(dif_spi_device_init_handle(
+      mmio_region_from_addr(TOP_EARLGREY_SPI_DEVICE_BASE_ADDR), &spi_device));
+
+  CHECK_DIF_OK(dif_rv_plic_init(
+      mmio_region_from_addr(TOP_EARLGREY_RV_PLIC_BASE_ADDR), &plic));
+
+  // Set IoA7 for tpm csb.
+  // Longer term this needs to migrate to a top specific, platform specific
+  // setting.
+  CHECK_DIF_OK(dif_pinmux_input_select(
+      &pinmux, kTopEarlgreyPinmuxPeripheralInSpiDeviceTpmCsb,
+      kTopEarlgreyPinmuxInselIoa7));
+
+  CHECK_DIF_OK(
+      dif_spi_device_tpm_configure(&spi_device, kDifToggleEnabled, tpm_config));
+
+  // enable interrupts
+  en_plic_irqs(&plic);
+  en_spi_device_irqs(&spi_device.dev);
+
+  // Sync message with testbench to begin.
+  LOG_INFO("Begin TPM Test");
+
+  for (uint32_t i = 0; i < kIterations; i++) {
+    LOG_INFO("Iteration %d", i);
+
+    // Wait for write interrupt.
+    while (!header_interrupt_received) {
+      wait_for_interrupt();
+    }
+    header_interrupt_received = false;
+
+    // Check what comamnd we have received. Store it as expected variables
+    // and compare when the read command is issued.
+    uint8_t write_command;
+    uint32_t write_addr;
+    CHECK_DIF_OK(dif_spi_device_tpm_get_command(&spi_device, &write_command,
+                                                &write_addr));
+    CHECK((write_command & kTpmCommandRwMask) == kTpmWriteCommand,
+          "Received write command is not expected");
+
+    // Poll for write data to complete.
+    uint32_t num_bytes = (write_command & kTpmCommandSizeMask) + 1;
+    LOG_INFO("Expecting %d bytes from tpm write", num_bytes);
+
+    uint8_t buf[64];
+    dif_result_t status = kDifOutOfRange;
+    while (status == kDifOutOfRange) {
+      status = dif_spi_device_tpm_read_data(&spi_device, num_bytes, buf);
+    };
+    CHECK_DIF_OK(status);
+
+    // Send the written data right back out for reads.
+    CHECK_DIF_OK(dif_spi_device_tpm_write_data(&spi_device, num_bytes, buf));
```

Comment:
Should we move this to the end of the loop? In reality, SW needs to get the request and know what address and size are in the request, before it writes the data to the fifo.

Reply:
hmmm... that's interesting, let me try that.